### ISSUE TITLE
proposal of a more functional file naming specification

### ISSUE THREAD

**Issue by vigilancetech-com** [protesilaos/denote#120]
Hi Prot:

I've been checking denote out for a  couple days now and something dawned on me.

Your date/time stamp has the letter "T" in between the date and the time.   IMO that's a wasted character.

Since the date and time fields are fixed length, such that they could be "cut" apart by character count, what I would propose as a more useful character there would be a number, say either 0-6 or 1-7, for the day of the week (e.g. 0=Sunday, 1=Monday, etc...).

That way when one is using dired to search for notes they could whip up a quick regexp to say "show me all the notes done on Wednesdays" for example.

Just a thought!

Thanks for your great work on this package!

**Comment by vigilancetech-com** [comment]
As another option maybe you could put a letter representing the day of the week so it would still set off the date stamp from the time stamp, e.g. S,M,T,W,H,F,A (where the second letter of the day name is used when the first is already taken).

But I think I still like the numeric digit better, like say if it's a zero based scale one could look for all the notes created on a weekday with a regexp:

^.\{6}[1-5]

**Comment by protesilaos** [comment]
Hello @vigilancetech-com!  This is a good idea.  I am not sure how to get the name of the current day and then ensure that the names are in English.  For example, in French there is more than one day that starts with "M" (Mardi, Mercredi) so I do not know what happens if someone has configuring their environment accordingly.

**Comment by protesilaos** [comment]
About using a number for the day of the week, I think that makes it more difficult to read the date and time components.  The `T` is easy to spot whereas, say, an extra `0` is not.

**Comment by sthesing** [comment]
> Your date/time stamp has the letter "T" in between the date and the time. IMO that's a wasted character.
> 
> Since the date and time fields are fixed length, such that they could be "cut" apart by character count, what I would propose as a more useful character there would be a number, say either 0-6 or 1-7, for the day of the week (e.g. 0=Sunday, 1=Monday, etc...).

While I like the thought to do something clever with the separator between date and time, I have two sceptical thoughts about this:
1. I agree whith Prot's argument for readability. While computers might be easily able to spot the border between sets of digits, humans have a harder time. I don't consider the T a wasted character. But, granted, another non-numerical character would do the job just as well.
2. So maybe use the letters A-G for days of the week? That would enable search operations on days (as @vigilancetech-com suggested), while both circumventing the doubling of letters (as in Tuesday-Thursday, Mardi-Mercredi etc.) and be language agnostic.

**Comment by nbehrnd** [comment]
> Your date/time stamp has the letter "T" in between the date and the time. IMO that's a wasted character.

The use of `T` as a separator between date and time equally is seen e.g., in Karl Voit's [date2time](https://github.com/novoid/date2name), a Python based implementation to add date/time stamps to files, run with either flag `--withtime` or the shorter `-w`.  Because `date2time` and `denote` are conceptual similar here, and because `denote` equally offers to annotate files, I would welcome if one could use either one application to yield the same result -- regardless if the engine relies on Emacs, or Python.  For some, the implementation into Windows' context menu `date2name` offers may be easier to engage, for other, living in Emacs is normal; then, exchange between colleagues of both worlds is eased.

> say either 0-6 or 1-7, for the day of the week (e.g. 0=Sunday, 1=Monday, etc...)
Regardless of the choice eventually adopted, I suggest to include an explicit note about this in the documentation.  European calendars typically start with Monday as first day of the 7 day week, while in the US and Canada it is Sunday instead.  As seen on the desks of colleagues, Algeria: it is Saturday, and Israel: Friday.  I don't know if either a) a manual adjustment by the user of `denote` is easier to implement, or b) a read-out of the OS system's locale to adjust this "under the hood".

**Comment by protesilaos** [comment]
I like this idea though it seems very tricky in practice.  We need to account for locales and somehow differentiate days that start with the same letter.

The proposal by @sthesing to use A-G for this case makes sense, though it may not feel natural.  Per the constraint with localisation, we will need to know which day is the first one.

This sort of work probably belongs to another package.  It might be able to leverage the built-in `calendar.el` to perform all the relevant calculations.  Then, it will provide the alternative identifier scheme through commands that alter the relevant functions we have to derive the file path and identifier.  On the `denote.el` front, we can accommodate such an extension by modifying the pattern of the identifier and/or supplying additional functions or points of entry.

Otherwise I cannot see this being done in `denote.el`: it likely is not a small feature.

**Comment by vigilancetech-com** [comment]
I would just say that the locale issue only becomes a concern if people are going to be sharing files with others that do it differently.   As this seems to be mostly a "personal database" at this time (or at least it seems so to me) that may not be such a big issue so it probably should be implemented, e.g. as A-G and let whomever is using it supply the meaning they want as to which day of the week it starts at (perhaps with a constant set in one's emacs config file, defaulting to the most common, which is probably the USA version, but I'm biased :) ) then sort out the locale issue later when/if it becomes a problem.

**Comment by protesilaos** [comment]
@vigilancetech-com

>  I would just say that the locale issue only becomes a concern if people are going to be sharing files with others that do it differently.

I am aware of at least three cases where this happens.  I also do it myself.  We cannot assume how users will organise their workflow.  The file-naming scheme of Denote does not rely on external software and thus makes files easier to share with others.  Even someone with no knowledge of the technicalities can infer some meaning from the file name alone.

As I noted before, I am concerned that the complex solution cannot be done in `denote.el`.  There are all sorts of edge cases here and we will be dealing with a lot of work that is not consistent with Denote's original design.  This adds to the maintenance burden and will likely lead to unforeseen issues that I would rather avoid.

* * *

An alternative here that will work for everyone and does not change Denote's file-naming scheme is to provide the user option `denote-auto-add-day-keyword`.  When set to `t`, all new notes will automatically add the name of the day to the file name.

This way we do not introduce a potentially confusing letter in the identifier and we make it easy to search for files taken on a given day.  It is more human-readable and may not require as much engineering work.

Even so, this is not trivial: we still need to have a way to figure out the name of the day and we still need to support localisation.  I am just mentioning it as an idea for us to consider.

**Comment by vigilancetech-com** [comment]
Maybe there should be an option for the user to select Zulu time for his file time stamps (if, for example, he plans on sharing his files, or moving them to a different time zone), then a letter like [A-G] in the date/timestamp file name would work for everyone, with just some tweaking to the searching routines to adjust for the locale (of that former option was set)?

**Comment by leinfink** [comment]
I just wanted to add that the T is also, to my knowledge, currently required in ISO 8601. To be honest, I thought that was the original motivation behind it for Denote.

**Comment by protesilaos** [comment]
> From: Henrik ***@***.***>
> Date: Fri,  3 Mar 2023 07:29:09 -0800
>
> I just wanted to add that the T is also, to my knowledge, currently
> required in ISO 8601. To be honest, I thought that was the original
> motivation behind it for Denote.

Indeed, it is inspired by the standard though it is not necessary for
file names as the operating system does not require/enforce ISO 8601.

The standard is used by Markdown files per the function
'denote-date-iso-8601'.  We do that because Markdown is used by static
site generators which require ISO 8601.

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by protesilaos** [comment]
Folks, I am closing this because it would add to the maintenance burden and is tricky to get right.
